students/rod_musser/lesson07/html_render.py

```diff
@@ -0,0 +1,130 @@
+#!/usr/bin/env python3
+
+"""
+A class-based system for rendering html.
+"""
+
+
+# This is the framework for the base class
+class Element(object):
+    tag_name = "html"
+    indent = "    "
+
+    def __init__(self, content=None, **kwargs):
+        if content is None:
+            self.content_list = []
+        else:
+            self.content_list = [content]
+        self.attributes = kwargs
+
+    def append(self, new_content):
+        self.content_list.append(new_content)
+
+    def render(self, out_file, curr_ident=""):
+        out_file.write(curr_ident + "<" + self.tag_name + render_attributes(self.attributes))
+        if isinstance(self, SelfClosingTag):
+            out_file.write(" />\n")
+        else:
+            out_file.write(">")
+            if not isinstance(self, OneLineTag):
+                out_file.write("\n")
+
+            for content in self.content_list:
+                if hasattr(content, 'render'):
+                    content.render(out_file, curr_ident + self.indent)
+                else:
+                    if not isinstance(self, OneLineTag):
+                        out_file.write(curr_ident + self.indent)
+                    out_file.write(str(content))
+                    if not isinstance(self, OneLineTag) and content == self.content_list[-1]:
+                        out_file.write("\n")
+            if not isinstance(self, OneLineTag):
+                out_file.write(curr_ident)
+            out_file.write("</" + self.tag_name + ">" + "\n")
```

Comment:
this is your base class - it should have no knowledge of the child classes 

Think about it this way, imagine you have a base car model then you can take that base model and do various changes or upgrades like color, sunroof etc. This is the same pattern, your base class is bare with common functionality then you can take it (think like extend it) and customize as needed.